### PR TITLE
Update `persistent_workers` recommendation when using spawn launcher

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Follow up to https://github.com/Lightning-AI/lightning/pull/18591

TODO: Provide examples and numbers showing the warnings don't apply anymore.
TODO: Remove it from docs: [lightning.ai/docs/pytorch/stable/advanced/speed.html#persistent-workers](https://lightning.ai/docs/pytorch/stable/advanced/speed.html#persistent-workers)

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: docs/source-pytorch/advanced/speed.rst [hunk 1]
```diff
@@ -101,26 +101,12 @@ For debugging purposes or for dataloaders that load very small datasets, it is d
 
     warnings.filterwarnings("ignore", category=PossibleUserWarning)
 
-Spawn
```
Comment (awaelchli) [inline docs/source-pytorch/advanced/speed.rst:104]: The trainer uses ddp (non-spawn) by default and there is no benefit of using ddp-spawn for speed. If you do use ddp-spawn, the only relevant consideration is the "persistent workers" setting explained in the section below.

File: src/lightning/pytorch/trainer/connectors/data_connector.py [hunk 2]
```diff
@@ -420,28 +420,17 @@ def _check_dataloader_iterable(
...
-                " may result in data loading bottlenecks."
-                " Consider setting persistent_workers=True"
-                " (this is a limitation of Python .spawn() and PyTorch)"
-            )
-
-    elif dataloader.num_workers == 0 and using_spawn:
-        if not dataloader.persistent_workers:
-            rank_zero_warn(
-                "strategy=ddp_spawn and num_workers=0 may result in data loading bottlenecks."
-                " Consider setting num_workers>0 and persistent_workers=True"
```
Comment (awaelchli) [inline src/lightning/pytorch/trainer/connectors/data_connector.py:443]: ddp-spawn is irrelevant here. The two other warnings already cover this one.

File: src/lightning/pytorch/trainer/connectors/data_connector.py [hunk 3]
```diff
@@ -420,28 +420,17 @@ def _check_dataloader_iterable(
         )
 
 
-def _worker_check(trainer: "pl.Trainer", using_spawn: bool, dataloader: object, name: str) -> None:
+def _worker_check(trainer: "pl.Trainer", dataloader: object, name: str) -> None:
     if not isinstance(dataloader, DataLoader):
         return
 
     upper_bound = suggested_max_num_workers(trainer.num_devices)
+    using_spawn = mp.get_start_method() == "spawn"
```
Comment (awaelchli) [inline src/lightning/pytorch/trainer/connectors/data_connector.py:428]: Since spawn is way more expensive than fork, this is why you notice the overhead so much and `persistent_workers=True` can help. 

On Linux, default is fork and on MacOs and Windows it is spawn. So these two platforms generally benefit from the persistent_workers settings while on Linux it is not noticeable. However, once the start method is set, you can't change it, meaning if we launch multi-GPU with ddp-spawn, even on Linux the dataloaders will use spawn. That's why the warning was initially tied to ddp-spawn but it is really just a consequence of the way `torch.multiprocessing` configures itself.

There is so much confusing and noisy information out there in forums etc, when it is really quite simple.

Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/data_connector.py:428]: I'm slightly worried that `mp.get_start_method` wont return the proper start method for unconventional strategies (xla, habana, ...?). It might be safer to keep accessing the strategy for this

File: tests/tests_pytorch/trainer/connectors/test_data_connector.py [hunk 4]
```diff
@@ -110,43 +108,37 @@ def test_dataloader(self):
...
+        self.warning_expected = warning_expected
 
     def on_fit_start(self):
-        self._resout = StringIO()
-        self.ctx = redirect_stderr(self._resout)
-        self.ctx.__enter__()
+        if self.warning_expected:
+            self.ctx = pytest.warns(UserWarning, match="Consider setting `persistent_workers=True`")
+        else:
+            self.ctx = no_warning_call(UserWarning, match="Consider setting `persistent_workers=True`")
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/connectors/test_data_connector.py:119]:
```suggestion
        if self.warning_expected:
            self.ctx = pytest.warns
        else:
            self.ctx = no_warning_call
        self.ctx = ctx(UserWarning, match="Consider setting `persistent_workers=True`")
```

**Review by carmocca — APPROVED** [review]
LGTM

File: src/lightning/pytorch/trainer/connectors/data_connector.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning/pytorch/trainer/connectors/data_connector.py]: Hmm that's ultimately how the DataLoader decides how to launch its workers. The dataloader also has a multiprocessing_context argument, we can take that into account: https://github.com/Lightning-AI/lightning/pull/18649/commits/400bd88f8d7f35c33a72275e8eb38ea678c4cf48

Note that our worker check here runs when processes have already been launched. Furthermore, it happens after the dataloader was processed by the strategy and for the XLA strategy, it is a [MpDeviceLoader](https://github.com/pytorch/xla/blob/1b68776e24ed90a5b48edbfca1e85b06128ab625/torch_xla/distributed/parallel_loader.py#L186)  which isn't a subclass of DataLoader and so we wouldn't emit a warning anyway.

**Review by Borda — APPROVED** [review]
